apps/small_filters/process.h

```diff
@@ -0,0 +1,526 @@
+
+    bool verify(const int W, const int H) {
+        const int16_t coeffs[5] = { 1, 4, 6, 4, 1 };
+        u8_out.for_each_element([&](int x, int y) {
+            int16_t blur = 0;
+            for (int rx = -2; rx < 3; ++rx) {
+                int16_t blur_y = 0;
+                for (int ry = -2; ry < 3; ++ry) {
+                    int16_t val = static_cast<int16_t>(u8_in(clamp(x+rx, 0, W-1), clamp(y+ry, 0, H-1)));
+                    blur_y += val * coeffs[ry + 2];
+                }
+                blur += blur_y * coeffs[rx + 2];
+            }
+            uint8_t blur_val = blur >> 8;
+            uint8_t out_xy = u8_out(x, y);
+            if (blur_val != out_xy) {
+                printf("Gaussian5x5: Mismatch at %d %d : %d != %d\n", x, y, out_xy, blur_val);
+                abort();
+            }
+        });
+        return true;
+    }
+
+    int run(bmark_run_mode_t mode) {
+        if (mode == bmark_run_mode_t::hvx64) {
+            return pipeline_64(u8_in, u8_out);
+        } else if (mode == bmark_run_mode_t::hvx128) {
+            return pipeline_128(u8_in, u8_out);
+        } else if (mode == bmark_run_mode_t::cpu) {
+            return pipeline_cpu(u8_in, u8_out);
+        }
+        abort();
+    }
+};
+
+class SobelDescriptor : public PipelineDescriptor<pipeline2> {
+    Halide::Runtime::Buffer<uint8_t> u8_in, u8_out;
+
+ public:
+ SobelDescriptor(pipeline2 pipeline_64, pipeline2 pipeline_128, pipeline2 pipeline_cpu,
+                     int W, int H) :
+    PipelineDescriptor<pipeline2>(pipeline_64, pipeline_128, pipeline_cpu), u8_in(nullptr, W, H, 2),
+        u8_out(nullptr, W, H, 2) {}
+
+    void init() {
+        u8_in.device_malloc(halide_hexagon_device_interface());
+        u8_out.device_malloc(halide_hexagon_device_interface());
+
+        u8_in.for_each_value([&](uint8_t &x) {
+                x = static_cast<uint8_t>(rand());
+        });
+        u8_out.for_each_value([&](uint8_t &x) {
+            x = 0;
+        });
+    }
+
+    const char *name() { return "sobel"; };
+    uint16_t sobel3(uint16_t a, uint16_t b, uint16_t c) {
+        return (a + 2*b + c);
+    }
+    bool verify(const int W, const int H) {
+        u8_out.for_each_element([&](int x, int y) {
+            uint16_t sobel_x_avg0 = sobel3(static_cast<uint16_t>(u8_in(clamp(x-1, 0, W-1), clamp(y-1, 0, H-1))),
+                                           static_cast<uint16_t>(u8_in(clamp(x, 0, W-1), clamp(y-1, 0, H-1))),
+                                           static_cast<uint16_t>(u8_in(clamp(x+1, 0, W-1), clamp(y-1, 0, H-1))));
+            uint16_t sobel_x_avg1 = sobel3(static_cast<uint16_t>(u8_in(clamp(x-1, 0, W-1), clamp(y+1, 0, H-1))),
+                                           static_cast<uint16_t>(u8_in(clamp(x, 0, W-1), clamp(y+1, 0, H-1))),
+                                           static_cast<uint16_t>(u8_in(clamp(x+1, 0, W-1), clamp(y+1, 0, H-1))));
+            uint16_t sobel_x = abs(sobel_x_avg0 - sobel_x_avg1);
```

Comment:
Does abs work on unsigned values like this? Does sobel3 need to return a signed int16_t?